Refactor documentation and add project overview sections (#19)
eProsima/DDS-Recorder: docs/rst/recording/usage/configuration.rst

    :depth: 2

DDS Recorder Configuration
==========================

A |ddsrecorder| is configured by a *.yaml* configuration file.

> This *.yaml* file contains all the information regarding the DDS interface configuration, recording parameters, and |ddsrecorder| specifiations.

Reviewer: ```suggestion
This *.yaml* file contains all the information regarding the DDS interface configuration, recording parameters, and |ddsrecorder| specifications.
```

Thus, this file has four major configuration groups:

* ``dds``: configuration related to DDS communication.
* ``recorder``: configuration of data writing in the database.

> * ``remote-controller``: configuration of the remote controller of the DDS Recorder.

Reviewer: ```suggestion
* ``remote-controller``: configuration of the remote controller of the |ddsrecorder|.
```

> * ``specs``: configuration of the internal operation of the DDS Recorder.

Reviewer: ```suggestion
* ``specs``: configuration of the internal operation of the |ddsrecorder|.
```

This way, it is possible to define a set of rules in |ddsrecorder| to filter those data samples the user does not wish to save.

It is not mandatory to define such set of rules in the configuration file.
In this case, a |ddsrecorder| will save all the data published under the topics that it automatically discovers within the DDS network to which it connects.

To define these data filtering rules based on the Topics to which they belong, the following lists are available:

* Allowed topics list (``allowlist``)
* Block topics list (``blocklist``)

These lists of topics stated above are defined by a tag in the *YAML* configuration file, which defines a *YAML* vector (``[]``).
This vector contains the list of topics for each filtering rule.
Each Topic is determined by its entries ``name``, ``type`` and ``keyed``, with only the first one being mandatory.

.. list-table::
    :header-rows: 1

    *   - Topic entries
        - Data type
        - Default value

    *   - ``name``
        - ``string``
        - \-

    *   - ``type``
        - ``string``
        - ``"*"``

    *   - ``keyed``
        - ``bool``

>         - Both ``true`` and ``false``

Reviewer: By default?

**Example of usage - Allowlist and blocklist collision:**

    In the following example, the ``HelloWorldTopic`` topic is both in the ``allowlist`` and (implicitly) in the
    ``blocklist``, so according to the ``blocklist`` preference rule this topic is blocked.
    Moreover, only the topics present in the allowlist are relayed, regardless of whether more topics are dynamically
    discovered in the DDS network.
    In this case the forwarded topics are ``AllowedTopic1`` with data type ``Allowed``
    and ``AllowedTopic2`` regardless of its data type.

    .. code-block:: yaml

        allowlist:
        - name: AllowedTopic1
            type: Allowed
        - name: AllowedTopic2
            type: "*"
        - name: HelloWorldTopic
            type: HelloWorld

        blocklist:
        - name: "*"
            type: HelloWorld

Built-in Topics
^^^^^^^^^^^^^^^

Apart from the dynamic DDS topics discovered in the network, the discovery phase can be accelerated by using the builtin topic list (``builtin-topics``).

> By defining topics in this list, the DDS recorder will create the DataWriters and DataReaders in recorder initialization.

Reviewer: ```suggestion
By defining topics in this list, the |ddsrecorder| will create the DataWriters and DataReaders in recorder initialization.
```

        - ``.``

    *   - File name
        - ``filename``
        - Configure the name of the output file.
        - ``string``
        -

Downsampling
^^^^^^^^^^^^

Reduces the sampling rate of the received data by keeping the first sample and then all *n-th* samples after the first sample, where *n* is the value specified in ``downsampling``.

Buffer size
^^^^^^^^^^^

``buffer-size`` indicates the number of samples to be stored in the process memory before the dump to disk.
This avoids disk access each time a sample is received.
This parameter applies to each topic independently, that is, ``buffer-size`` samples of each topic will be buffered before writing them to the database.

Event Duration
^^^^^^^^^^^^^^

The |ddsrecorder| can be configured to continue saving data when it is in pause mode.
Thus, when the |ddsrecorder| receives the event from the remote controller, it will save the samples received in the time period configured in ``event-duration`` in the database and restart the timer.

In other words, the ``event-duration`` acts as a sliding time window that allows to save the collected samples in this time window only when the remote controller event is received.

> This way, a DDS Recorder is working in pause mode, recording samples in the database only when the event from the remote controller occurs.

Reviewer: ```suggestion
This way, a |ddsrecorder| is working in pause mode, recording samples in the database only when the event from the remote controller occurs.
```

-------------------

The internals of the |ddsrecorder| can be configured using the ``specs`` optional tag that contains certain options related with the overall configuration of the |ddsrecorder| instance to run.
The values available to configure are:

Number of Threads
^^^^^^^^^^^^^^^^^

``specs`` supports a ``threads`` optional value that allows the user to set a maximum number of threads for the internal :code:`ThreadPool`.
This ThreadPool allows to limit the number of threads spawned by the application.
This improves the performance of the internal data communications.

This value should be set by each user depending on each system characteristics.
In case this value is not set, the default number of threads used is :code:`12`.

Maximum Number of Pending Samples
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

It is possible that the |ddsrecorder| starts receiving data from a topic that it has not yet registered, i.e. a topic for which it does not know the data type.
In order not to discard the samples received from this topic, it is possible to keep a limited number of samples in an internal circular buffer that stores those samples that do not yet have a known data type.

> The max-pending-samples parameter allows to configure the size of this circular buffer for each topic that is discovered.

Reviewer: ```suggestion
The ``max-pending-samples`` parameter allows to configure the size of this circular buffer for each topic that is discovered.
```

The default value is equal to ``10`` samples.

.. _history_depth_configuration:

Maximum History Depth
^^^^^^^^^^^^^^^^^^^^^

> ``specs`` supports a ``max-depth`` **optional** value that configures the history size of the Fast DDS internal entities.

Reviewer: ```suggestion
``specs`` supports a ``max-depth`` **optional** value that configures the history size of the *Fast DDS* internal entities.
```

.. code-block:: yaml

    dds:

>     domain: 0

Reviewer: Indentation

    allowlist:
      - name: "topic_name"
        type: "topic_type"

    blocklist:
      - name: "topic_name"
        type: "topic_type"

    recorder:
      output:
        extension: ".mcap"
        filename: "output"
        path: "."

      downsampling: 0.1
      buffer-size: 50
      event-duration: 60

    remote-controller:
      enable: true
      domain: 10

    specs:
      threads: 8
      max-depth: 100
      max-pending-samples: 10

.. _usage_fastdds_configuration:

Fast DDS Configuration
======================

> As mentioned before, the DDS Recorder requires the topic types in order to be able to record the data of such topics.

Reviewer: ```suggestion
As mentioned before, the DDS Record requires the topic types in order to be able to record the data of such topics.
```

This requires that the user application needs to be configured to send the required type information.
However, *Fast DDS* does not send the data type information by default, it must be configured to do so.
First of all, when generating the topic types using *eProsima Fast DDS Gen*, the option ``-typeobject`` must be added in order to generate the needed code to fill the ``TypeObject`` data.

> For native types (data types that does not rely in other data types) this is enough, as *Fast DDS* will send the TypeObject by default.

Reviewer: ```suggestion
For native types (data types that does not rely in other data types) this is enough, as *Fast DDS* will send the ``TypeObject`` by default.
```
